// microsoft-partnercenter/dist/lib/types/subscriptions.types.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.UnitType = exports.TermDuration = exports.SuspensionReason = exports.Status = exports.ContractType = exports.BillingType = exports.BillingCycle = exports.Action = void 0;
var Action;
(function (Action) {
    Action["Cancel"] = "Cancel";
    Action["Edit"] = "Edit";
})(Action || (exports.Action = Action = {}));
var ObjectType;
(function (ObjectType) {
    ObjectType["Subscription"] = "Subscription";
})(ObjectType || (ObjectType = {}));
var BillingCycle;
(function (BillingCycle) {
    BillingCycle["Monthly"] = "monthly";
})(BillingCycle || (exports.BillingCycle = BillingCycle = {}));
var BillingType;
(function (BillingType) {
    BillingType["License"] = "license";
    BillingType["Usage"] = "usage";
})(BillingType || (exports.BillingType = BillingType = {}));
var ContractType;
(function (ContractType) {
    ContractType["Subscription"] = "subscription";
})(ContractType || (exports.ContractType = ContractType = {}));
var Status;
(function (Status) {
    Status["Active"] = "active";
    Status["Deleted"] = "deleted";
    Status["Disabled"] = "disabled";
    Status["Suspended"] = "suspended";
    Status["Expired"] = "expired";
})(Status || (exports.Status = Status = {}));
var SuspensionReason;
(function (SuspensionReason) {
    SuspensionReason["Expiration"] = "Expiration";
})(SuspensionReason || (exports.SuspensionReason = SuspensionReason = {}));
var TermDuration;
(function (TermDuration) {
    TermDuration["P1M"] = "P1M";
    TermDuration["P1Y"] = "P1Y";
})(TermDuration || (exports.TermDuration = TermDuration = {}));
var UnitType;
(function (UnitType) {
    UnitType["Licenses"] = "Licenses";
    UnitType["UsageBased"] = "Usage-based";
})(UnitType || (exports.UnitType = UnitType = {}));
