// react-microsoft-login-advanced/dist/MicrosoftLogin.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
Object.defineProperty(exports, "__esModule", { value: true });
var React = require("react");
var msal_browser_fixed_beta_1 = require("@dreaken/msal-browser-fixed-beta");
var MicrosoftLoginButton_1 = require("./MicrosoftLoginButton");
var CLIENT_ID_REGEX = /[a-z0-9]{8}-[a-z0-9]{4}-[a-z0-9]{4}-[a-z0-9]{4}-[a-z0-9]{12}/;
var getUserAgentApp = function (_a) {
    var clientId = _a.clientId, tenantUrl = _a.tenantUrl, redirectUri = _a.redirectUri, responseType = _a.responseType;
    if (clientId && CLIENT_ID_REGEX.test(clientId)) {
        return new msal_browser_fixed_beta_1.PublicClientApplication({
            auth: __assign(__assign(__assign(__assign({}, (redirectUri && { redirectUri: redirectUri })), (tenantUrl && { authority: tenantUrl })), (responseType && { responseType: responseType })), { clientId: clientId, navigateToLoginRequestUrl: false })
        });
    }
};
var getScopes = function (graphScopes) {
    return graphScopes || [];
};
var MicrosoftLogin = (function (_super) {
    __extends(MicrosoftLogin, _super);
    function MicrosoftLogin(props) {
        var _this = _super.call(this, props) || this;
        _this.login = function () {
            var _a = _this.state, msalInstance = _a.msalInstance, scopes = _a.scopes;
            var _b = _this.props, _c = _b.withUserData, withUserData = _c === void 0 ? false : _c, authCallback = _b.authCallback, _d = _b.forceRedirectStrategy, forceRedirectStrategy = _d === void 0 ? false : _d, prompt = _b.prompt, extraQueryParameters = _b.extraQueryParameters;
            if (msalInstance) {
                _this.log("Login STARTED");
                if (forceRedirectStrategy || _this.checkToIE()) {
                    _this.redirectLogin({
                        msalInstance: msalInstance,
                        scopes: scopes,
                        prompt: prompt,
                        extraQueryParameters: extraQueryParameters
                    });
                }
                else {
                    _this.popupLogin({
                        msalInstance: msalInstance,
                        scopes: scopes,
                        withUserData: withUserData,
                        authCallback: authCallback,
                        prompt: prompt,
                        extraQueryParameters: extraQueryParameters
                    });
                }
            }
            else {
                _this.log("Login FAILED", "clientID broken or not provided", true);
            }
        };
        var graphScopes = props.graphScopes, clientId = props.clientId, tenantUrl = props.tenantUrl, redirectUri = props.redirectUri, responseType = props.responseType, extraQueryParameters = props.extraQueryParameters;
        _this.state = {
            msalInstance: getUserAgentApp({ clientId: clientId, tenantUrl: tenantUrl, redirectUri: redirectUri, responseType: responseType }),
            scopes: getScopes(graphScopes),
            extraQueryParameters: extraQueryParameters
        };
        return _this;
    }
    MicrosoftLogin.prototype.componentDidMount = function () {
        var _this = this;
        var _a = this.state, msalInstance = _a.msalInstance, scopes = _a.scopes, extraQueryParameters = _a.extraQueryParameters;
        var _b = this.props, authCallback = _b.authCallback, _c = _b.withUserData, withUserData = _c === void 0 ? false : _c;
        if (!msalInstance) {
            this.log("Initialization", "clientID broken or not provided", true);
        }
        else {
            msalInstance.handleRedirectCallback(function (error, authResponse) {
                if (!error && authResponse) {
                    _this.log("Fetch Azure AD 'token' with redirect SUCCEDEED", authResponse);
                    authCallback(undefined, authResponse);
                }
                else {
                    _this.log("Fetch Azure AD 'token' with redirect FAILED", error, true);
                    authCallback(error);
                }
            });
        }
    };
    MicrosoftLogin.prototype.componentDidUpdate = function (prevProps) {
        var _a = this.props, clientId = _a.clientId, tenantUrl = _a.tenantUrl, redirectUri = _a.redirectUri;
        if (prevProps.clientId !== clientId ||
            prevProps.tenantUrl !== tenantUrl ||
            prevProps.redirectUri !== redirectUri) {
            this.setState({
                msalInstance: getUserAgentApp({ clientId: clientId, tenantUrl: tenantUrl, redirectUri: redirectUri })
            });
        }
    };
    MicrosoftLogin.prototype.popupLogin = function (_a) {
        var _this = this;
        var msalInstance = _a.msalInstance, scopes = _a.scopes, withUserData = _a.withUserData, authCallback = _a.authCallback, prompt = _a.prompt, extraQueryParameters = _a.extraQueryParameters;
        this.log("Fetch Azure AD 'token' with popup STARTED");
        msalInstance
            .loginPopup({ scopes: scopes, prompt: prompt, extraQueryParameters: extraQueryParameters })
            .then(function (authResponse) {
            _this.log("Fetch Azure AD 'token' with popup SUCCEDEED", authResponse);
            _this.log("Fetch Graph API 'access_token' in silent mode STARTED");
        })
            .catch(function (error) {
            _this.log("Fetch Azure AD 'token' with popup FAILED", error, true);
            authCallback(error);
        });
    };
    MicrosoftLogin.prototype.redirectLogin = function (_a) {
        var msalInstance = _a.msalInstance, scopes = _a.scopes, prompt = _a.prompt, extraQueryParameters = _a.extraQueryParameters;
        this.log("Fetch Azure AD 'token' with redirect STARTED");
        msalInstance.loginRedirect({ scopes: scopes, prompt: prompt, extraQueryParameters: extraQueryParameters });
    };
    MicrosoftLogin.prototype.checkToIE = function () {
        var ua = window.navigator.userAgent;
        var msie = ua.indexOf("MSIE ");
        var msie11 = ua.indexOf("Trident/");
        var msedge = ua.indexOf("Edge/");
        var isIE = msie > 0 || msie11 > 0;
        var isEdge = msedge > 0;
        return isIE || isEdge;
    };
    MicrosoftLogin.prototype.log = function (name, content, isError) {
        var debug = this.props.debug;
        if (debug) {
            var style = "background-color: " + (isError ? "#990000" : "#009900") + "; color: #ffffff; font-weight: 700; padding: 2px";
            console.groupCollapsed("MSLogin debug");
            console.log("%c" + name, style);
            content && console.log(content.message || content);
            console.groupEnd();
        }
    };
    MicrosoftLogin.prototype.render = function () {
        var _a = this.props, buttonTheme = _a.buttonTheme, className = _a.className, children = _a.children;
        return children ? (React.createElement("div", { onClick: this.login }, children)) : (React.createElement(MicrosoftLoginButton_1.default, { buttonTheme: buttonTheme || "light", buttonClassName: className, onClick: this.login }));
    };
    return MicrosoftLogin;
}(React.Component));
exports.default = MicrosoftLogin;
